community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/AccessModeSnapshot.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.security;
+
+import org.neo4j.graphdb.security.AuthorizationViolationException;
+import org.neo4j.kernel.api.security.AccessMode;
+
+public class AccessModeSnapshot implements AccessMode
+{
+    private final boolean allowsReads;
+    private final boolean allowsWrites;
+    private final boolean allowsSchemaWrites;
+
+    private final AccessMode originalMode;
+
+    public static AccessMode create( AccessMode accessMode )
+    {
+        // TODO: Use flyweight pattern instead of always creating new objects, when we have a proper
+        //       security context and do not need to obtain the original mode through this object
+        return new AccessModeSnapshot( accessMode );
```

Comment:
No, because the access mode could be an `AuthSubject` (it almost always is), and we need to be able to get hold of it by a downcast of the original access mode (until we can do the refactoring mentioned above). I guess we could still do a flyweight including the original access mode reference, but it would have much more combinations so it didn't seem worthwhile.